// @ant-design/use-emotion-css/dist/demo/base.js
import { useEmotionCss } from "./..";
import { Button } from 'antd';
import React from 'react';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
export default (function () {
  var className = useEmotionCss(function (_ref) {
    var token = _ref.token;
    return {
      color: token.colorPrimary,
      display: 'flex',
      justifyContent: 'center',
      gap: 24,
      padding: 24,
      flexDirection: 'column'
    };
  });
  var blockClassName = useEmotionCss(function (_ref2) {
    var token = _ref2.token;
    return {
      backgroundColor: token.colorPrimary,
      width: 12,
      height: 12,
      borderRadius: 12
    };
  });
  return /*#__PURE__*/_jsxs("div", {
    className: className,
    children: [/*#__PURE__*/_jsxs("div", {
      style: {
        display: 'flex',
        alignItems: 'center',
        gap: 8
      },
      children: [/*#__PURE__*/_jsx("div", {
        className: blockClassName
      }), /*#__PURE__*/_jsx("span", {
        children: "Hello World"
      })]
    }), /*#__PURE__*/_jsx("div", {
      children: /*#__PURE__*/_jsx(Button, {
        type: "primary",
        children: "\u6309\u94AE"
      })
    })]
  });
});
