// @umijs/plugins/dist/antd.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/antd.ts
var antd_exports = {};
__export(antd_exports, {
  default: () => antd_default
});
module.exports = __toCommonJS(antd_exports);
var import_assert = __toESM(require("assert"));
var import_path = require("path");
var import_umi = require("umi");
var import_plugin_utils = require("umi/plugin-utils");
var import_resolveProjectDep = require("./utils/resolveProjectDep");
var import_withTmpPath = require("./utils/withTmpPath");
var antd_default = (api) => {
  let pkgPath;
  let antdVersion = "4.0.0";
  try {
    pkgPath = (0, import_resolveProjectDep.resolveProjectDep)({
      pkg: api.pkg,
      cwd: api.cwd,
      dep: "antd"
    }) || (0, import_path.dirname)(require.resolve("antd/package.json"));
    antdVersion = require(`${pkgPath}/package.json`).version;
  } catch (e) {
  }
  api.describe({
    config: {
      schema(Joi) {
        return Joi.alternatives().try(
          Joi.object({
            configProvider: Joi.object(),
            // themes
            dark: Joi.boolean(),
            compact: Joi.boolean(),
            // babel-plugin-import
            import: Joi.boolean(),
            // less or css, default less
            style: Joi.string().allow("less", "css"),
            theme: Joi.object()
          }),
          Joi.boolean().invalid(true)
        );
      }
    },
    enableBy({ userConfig }) {
      return process.env.UMI_PLUGIN_ANTD_ENABLE || userConfig.antd;
    }
  });
  api.addRuntimePluginKey(() => ["antd"]);
  function checkPkgPath() {
    if (!pkgPath) {
      throw new Error(`Can't find antd package. Please install antd first.`);
    }
  }
  api.modifyAppData((memo) => {
    checkPkgPath();
    const version = require(`${pkgPath}/package.json`).version;
    memo.antd = {
      pkgPath,
      version
    };
    return memo;
  });
  api.modifyConfig((memo) => {
    var _a;
    checkPkgPath();
    let antd = memo.antd || {};
    if (process.env.UMI_PLUGIN_ANTD_ENABLE) {
      const { defaultConfig } = JSON.parse(process.env.UMI_PLUGIN_ANTD_ENABLE);
      memo.antd = antd = Object.assign(defaultConfig, antd);
    }
    memo.alias.antd = pkgPath;
    if (antd.dayjs) {
      memo.alias.moment = (0, import_path.dirname)(require.resolve("dayjs/package.json"));
    }
    if (antdVersion.startsWith("5")) {
      const theme = require("@ant-design/antd-theme-variable");
      memo.theme = {
        ...theme,
        ...memo.theme
      };
      if ((_a = memo.antd) == null ? void 0 : _a.import) {
        const errorMessage = `Can't set antd.import while using antd5 (${antdVersion})`;
        api.logger.fatal(
          "please change config antd.import to false, then start server again"
        );
        throw Error(errorMessage);
      }
    }
    if (antd.dark || antd.compact) {
      const { getThemeVariables } = require("antd/dist/theme");
      memo.theme = {
        ...getThemeVariables(antd),
        ...memo.theme
      };
    }
    memo.theme = {
      "root-entry-name": "default",
      ...memo.theme
    };
    if (antd.theme) {
      (0, import_assert.default)(
        antdVersion.startsWith("5"),
        `antd.theme is only valid when antd is 5`
      );
      antd.configProvider ?? (antd.configProvider = {});
      antd.configProvider.theme = (0, import_plugin_utils.deepmerge)(
        antd.configProvider.theme || {},
        antd.theme
      );
    }
    return memo;
  });
  api.addExtraBabelPlugins(() => {
    const style = antdVersion.startsWith("5") ? false : api.config.antd.style || "less";
    return api.config.antd.import && !api.appData.vite ? [
      [
        require.resolve("babel-plugin-import"),
        {
          libraryName: "antd",
          libraryDirectory: "es",
          ...style ? {
            style: style === "less" || "css"
          } : {}
        },
        "antd"
      ]
    ] : [];
  });
  api.onGenerateFiles(() => {
    if (!api.config.antd.configProvider)
      return;
    api.writeTmpFile({
      path: `runtime.tsx`,
      content: import_plugin_utils.Mustache.render(
        `
import React from 'react';
import { ConfigProvider, Modal, message, notification } from 'antd';
import { ApplyPluginsType } from 'umi';
import { getPluginManager } from '../core/plugin';

export function rootContainer(container) {
  const finalConfig = getPluginManager().applyPlugins({
    key: 'antd',
    type: ApplyPluginsType.modify,
    initialValue: {...{{{ config }}}},
  });
  if (finalConfig.prefixCls) {
    Modal.config({
      rootPrefixCls: finalConfig.prefixCls
    });
    message.config({
      prefixCls: \`\${finalConfig.prefixCls}-message\`
    });
    notification.config({
      prefixCls: \`\${finalConfig.prefixCls}-notification\`
    });
  }
  if (finalConfig.iconPrefixCls) {
    // Icons in message need to set iconPrefixCls via ConfigProvider.config()
    ConfigProvider.config({
      iconPrefixCls: finalConfig.iconPrefixCls,
    });
  }
  return <ConfigProvider {...finalConfig}>{container}</ConfigProvider>;
}
      `.trim(),
        {
          config: JSON.stringify(api.config.antd.configProvider)
        }
      )
    });
    api.writeTmpFile({
      path: "types.d.ts",
      content: `
import type { ConfigProviderProps } from 'antd/es/config-provider';
export type RuntimeAntdConfig = (memo: ConfigProviderProps) => ConfigProviderProps;
`
    });
    api.writeTmpFile({
      path: import_umi.RUNTIME_TYPE_FILE_NAME,
      content: `
import type { RuntimeAntdConfig } from './types.d';
export type IRuntimeConfig = {
  antd?: RuntimeAntdConfig
};
      `
    });
  });
  api.addRuntimePlugin(() => {
    return api.config.antd.configProvider ? [(0, import_withTmpPath.withTmpPath)({ api, path: "runtime.tsx" })] : [];
  });
  api.addEntryImportsAhead(() => {
    const style = api.config.antd.style || "less";
    const doNotImportLess = api.config.antd.import && !api.appData.vite || antdVersion.startsWith("5");
    return doNotImportLess ? [] : [
      {
        source: style === "less" ? "antd/dist/antd.less" : "antd/dist/antd.css"
      }
    ];
  });
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {});
