// umi-presets-pro/dist/features/maxtabs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.withTmpPath = exports.winPath = void 0;
const path_1 = require("path");
function winPath(path) {
    const isExtendedLengthPath = /^\\\\\?\\/.test(path);
    if (isExtendedLengthPath) {
        return path;
    }
    return path.replace(/\\/g, '/');
}
exports.winPath = winPath;
function withTmpPath(opts) {
    return winPath((0, path_1.join)(opts.api.paths.absTmpPath, opts.api.plugin.key && !opts.noPluginDir
        ? `plugin-${opts.api.plugin.key}`
        : '', opts.path));
}
exports.withTmpPath = withTmpPath;
exports.default = (api) => {
    var _a;
    if ((_a = api.userConfig) === null || _a === void 0 ? void 0 : _a.keepalive) {
        api.onGenerateFiles(() => {
            api.writeTmpFile({
                path: 'Layout.tsx',
                content: `
        import { useKeepOutlets } from '@umijs/max';
  
        const Layout = () => {
          const element = useKeepOutlets();
          return <>{element}</>;
        };
  
        export default Layout;
  `,
            });
        });
        api.register({
            key: 'addLayouts',
            fn() {
                return [
                    {
                        id: 'max-tabs',
                        file: withTmpPath({ api, path: 'Layout.tsx' }),
                        test: (route) => {
                            return route.layout !== false;
                        },
                    },
                ];
            },
            stage: -1,
        });
    }
};
